src/Functions/stringToMap.cpp

```diff
@@ -0,0 +1,278 @@
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        size_t num_arguments = arguments.size();
+
+        if (num_arguments < 1 || num_arguments > 3)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Number of arguments for function {} doesn't match: "
+                            "passed {}, should be 1, 2 or 3", getName(), num_arguments);
+        if (!isStringOrFixedString(arguments[0]))
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}",
+                            arguments[0]->getName(), getName());
+        if (num_arguments >= 2 && !isStringOrFixedString(arguments[1]))
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}",
+                            arguments[1]->getName(), getName());
+        if (num_arguments == 3 && !isStringOrFixedString(arguments[2]))
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}",
+                            arguments[2]->getName(), getName());
+
+        DataTypePtr string_type = std::make_shared<DataTypeString>();
+        DataTypePtr string_type_nullable = std::make_shared<DataTypeNullable>(string_type);
+        DataTypes key_value_type{string_type, string_type_nullable};
+
+        return std::make_shared<DataTypeMap>(key_value_type);
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    static bool tryExecuteString(const IColumn * col, ColumnPtr & col_res, char pair_delim, char key_value_delim)
+    {
+        const ColumnString * col_str_in = checkAndGetColumn<ColumnString>(col);
+
+        if (col_str_in)
+        {
+            const ColumnString::Chars & in_vec = col_str_in->getChars();
+            const ColumnString::Offsets & in_offsets = col_str_in->getOffsets();
+
+            size_t size = in_offsets.size();
+
+            MutableColumnPtr keys_data = DataTypeString().createColumn();
+            MutableColumnPtr values_data = DataTypeString().createColumn();
+            MutableColumnPtr offsets = DataTypeNumber<IColumn::Offset>().createColumn();
+
+            ColumnUInt8::MutablePtr values_null_map = ColumnUInt8::create();
+
+            ColumnString::Offset curr_res_offset = 0, curr_in_offset = 0;
+
+            for (size_t i = 0; i < size; ++i)
+            {
+                const char * pos_in = reinterpret_cast<const char *>(&in_vec[curr_in_offset]);
+                size_t curr_size = strlen(pos_in);
+                const char * pos_in_end = pos_in + curr_size;
+                const char * pos_delim = nullptr;
+                std::vector<String> key_value_pairs;
+                // splitting the string into key-value pairs
+                while (pos_delim != pos_in_end)
+                {
+                    pos_delim = reinterpret_cast<const char *>(strchr(pos_in, pair_delim));
```

Comment:
Let's do something more "modern", e.g. wrap the input string in a `std::string_view` and call `.find()` on it.

Or just use `boost::algorithm::split()`.